### PR TITLE
Add RasterCache metrics to the FrameTimings

### PR DESCRIPTION
**Pull request by flar**
Fixes https://github.com/flutter/flutter/issues/89402

Currently RasterCache metrics are reported on every frame via the timeline trace mechanism. The non e2e benchmarks which get their data from those traces will soon [collect those metrics out of the timeline traces](https://github.com/flutter/flutter/pull/89393) and report them as trackable metrics to skia perf. Unfortunately, the e2e benchmarks use this FrameTimings mechanism to get their data and so can not see those timeline trace events unless they enable the potentially expensive timeline tracing. This PR adds the RasterCache metrics to the FrameTiming reporting so that both mechanisms can now report cache metrics as trackable data in the perf database.

The sweep count is an attempt to make sure we call these metrics under a consistent time with respect to the cache being swept. In this case we are making sure there are no sweeps between RasterStart and RasterEnd events.

This is logically different than the way that the metrics are handled in the timeline traces where the RasterCache event is posted after the caches are swept, at the end of the `RasterCache::SweepAfterFrame` method which is how it has been done historically.

In order for the FrameTimings mechanism to report the sweep metrics the same way as the timeline events, it would have to allow the sweep before it finalized the timings for the frame, which is currently done at RasterEnd time. Unfortunately, the cache sweeping is handled outside a scope where the RasterEnd time is recorded so that is difficult. Also, it isn't clear that we'd want to wait for the caches to be swept before we record the raster end time (do we?).

Possible changes to bring the 2 mechanisms into line:

- Have RasterEnd call sweep after it records the times, but before it updates the metrics in the appropriate fields. This would then take the responsibility away from the current agent that does this (`CompositorContext::EndFrame`) and would mean that the event would be added to the timeline from the `FrameTimings::RecordRasterEnd` call. The metrics reported by both mechanisms would then be post-sweep.
- Create another stage to the state diagram of the FrameTimingsRecorder that would allow the cache metrics to be added after the raster end was recorded. The metrics could then be added before or after the sweep is done. The downside is that it adds another stage of potential state violations to the existing sequence of state transitions in the Recorder.
- [This option is implemented in the latest commits] Switch to all reporting systems using pre-sweep metrics for the cache. This would provide a more accurate assessment of how much cache memory was consumed (but not necessarily used) during the frame because the old entries are currently held until after we've rendered the whole frame even though they aren't being used in that frame. In a very real sense they were a memory pressure burden while that frame was being rendered and so they should be reported in some manner. Currently we hide that pressure and literally "Sweep" it under a rug before we report the metrics. The proposed implementation here will do this "pre-sweep" reporting for the FrameTimings but the Trace Events are still emitted after the caches are swept (see `RasterCache::SweepAfterFrame`) - consistency can be achieved here by simply moving the trace event call to before the caches are swept (move it up 5 lines within that same method).

Note that we already have issues filed about the ways in which we mismanage the cache memory so this mechanism is geared towards providing a good reporting of the current state of affairs for now, not the future ideal.

### CODE REVIEW

**Comment by flar** [comment]
One big TBD here - get the new data to the Dart version of the class.

**Comment by flar** [comment]
The latest commit has the following changes:

- I upload the data to Dart now. I have a patch ready to go to have the e2e benchmarks in the framework expose this data to skia perf.
- I moved the call to emit the trace events for the cache to before the cache is swept for consistency with the FrameTimings statistics. The amount of memory reported for gallery increased a bit as a result, but the e2e and timeline versions of the benchmarks now collect very similar raster cache metrics to each other.

**Comment by flar** [comment]
@dnfield @chinmaygarde @iskakaushik

**Comment by flar** [comment]
I've just confirmed (what should be fairly obvious, but just to confirm) that the extra data being collected and transferred to the Dart FrameTiming objects does not interfere with the unmodified e2e benchmarks in the framework so this PR should be safe to land and roll before we add the framework-side support to publish the metrics to skia perf.

**Comment by flar** [comment]
My impression, in modifying the web_ui version of FrameTimings, was that web does not use layer or picture caching so its cache metrics would all just default to 0.

@yjbanov ?

File: flow/frame_timings.cc [hunk 1]
```diff
@@ -102,26 +130,41 @@ void FrameTimingsRecorder::RecordBuildEnd(fml::TimePoint build_end) {
...
+        picture_cache_bytes_ = 0;
+  }
   timing_.Set(FrameTiming::kVsyncStart, vsync_start_);
   timing_.Set(FrameTiming::kBuildStart, build_start_);
   timing_.Set(FrameTiming::kBuildFinish, build_end_);
   timing_.Set(FrameTiming::kRasterStart, raster_start_);
   timing_.Set(FrameTiming::kRasterFinish, raster_end_);
   timing_.Set(FrameTiming::kRasterFinishWallTime, raster_end_wall_time_);
   timing_.SetFrameNumber(GetFrameNumber());
+  timing_.SetCacheStatistics(layer_cache_count_, layer_cache_bytes_,
```
Comment (iskakaushik) [inline flow/frame_timings.cc:166]: nit: Consider `SetRasterCacheStatistics`

File: lib/ui/platform_dispatcher.dart [hunk 2]
```diff
@@ -1273,16 +1302,54 @@ class FrameTiming {
...
+  ///
+  /// See also [layerCacheCount], [layerCacheBytes], [pictureCacheCount] and [pictureCacheBytes].
+  double get pictureCacheMegabytes => pictureCacheBytes / 1024.0 / 1024.0;
+
   /// The frame key associated with this frame measurement.
-  int get frameNumber => _timestamps.last;
+  int get frameNumber => _data.last;
 
-  final List<int> _timestamps;  // in microseconds
+  final List<int> _data;  // in microseconds
```
Comment (iskakaushik) [inline lib/ui/platform_dispatcher.dart:1338]: it's no longer in microseconds :-)

**Review by iskakaushik — APPROVED** [review]
LGTM modulo the 2 nits.

File: flow/frame_timings.cc (hunk 1 above)
Comment (flar) [reply flow/frame_timings.cc]: Done

File: lib/ui/platform_dispatcher.dart (hunk 2 above)
Comment (flar) [reply lib/ui/platform_dispatcher.dart]: Done

**Pull request merged**